python_benchmarks/core.py

```diff
@@ -0,0 +1,202 @@
+L2_CACHE_SIZE, PEAK_BANDWIDTH = get_device_properties()
+
+
+def clear_l2_cache() -> None:
+    """
+    Flushes the L2 cache by creating a buffer of the same size.
+    """
+    n_elements = L2_CACHE_SIZE // 4
+    x = torch.empty(n_elements, dtype=torch.float32, device="cuda", requires_grad=False)
+    y = torch.clone(x)
+
+
+class NVFBenchmark:
+    """
+    A wrapper class around pytest-benchmark to support
+    torchprofiler-based timer and metric computation.
+    """
+
+    def __init__(self, benchmark_fixture, precision: float = 1e-6):
+        """
+        Arguments:
+            benchmark_fixture: pytest-benchmark fixture passed to every
+                function intended to be run as a benchmark by pytest
+            precision: Precision for the torchprofiler-based timer used.
+                Set explicitly to avoid timer calibration.
+
+        Class members:
+            self.prof: torch.profiler instance used as by the custom torchprofile_timer for the current benchmark
+            self.benchmark: Underlying pytest-benchmark fixture with timer modified to use torchprofile_timer
+            self.current_time: Global montonic clock incremented based on elapsed CUDA time
+        """
+        # Initialize a Torch Profiler object
+        self.prof = profile(activities=[ProfilerActivity.CUDA, ProfilerActivity.CPU])
+
+        # Modify the default timer.
+        benchmark_fixture._timer = self.torchprofile_timer
+        # Externally set the precision to avoid timer calibration. Since the timer uses CUDA times,
+        # calibration using subsequent timer calls produces invalid results.
+        # https://github.com/ionelmc/pytest-benchmark/blob/728752d2976ef53fde7e40beb3e55f09cf4d4736/src/pytest_benchmark/timers.py#L15
+        benchmark_fixture._precisions[benchmark_fixture._timer] = precision
+
+        self.benchmark = benchmark_fixture
+
+        # Global montonic clock
+        self.current_time = 0.0
+
+    def __call__(self, function_to_benchmark: Callable, *args, **kwargs):
+        return self.benchmark(function_to_benchmark, *args, **kwargs)
+
+    def __getattr__(self, attr):
+        if attr not in self.__dict__:
+            return getattr(self.benchmark, attr)
+        return super().__getattr__(attr)
+
+    def torchprofile_timer(self) -> float:
+        """
+        Custom torchprofiler-based timer used by pytest-benchmark.
+        At every timer call, the profiler is stopped to compute the elapsed CUDA time
+        and the global clock is incremented. The profiler is restarted before returning to continue tracing.
+
+        Returns:
+            self.current_time: Global monotonic clock variable
+        """
+        try:
+            self.prof.stop()
+            prof_averages = self.prof.key_averages()
+            elapsed_cuda_time = self._get_kernel_time(prof_averages)
+            self._increment_global_time(elapsed_cuda_time)
+            self.prof.start()
+        except:
+            self.prof.start()
```

Comment:
What exceptions could be thrown here by `self.prof.stop()`? Can we narrow which ones we catch so that we don't silently reset the timer? If we keep the try, let's describe why in a comment here.

Reply:
If the profiler has not been started or already stopped, it throws an `AssertionError`. Changed it to specifically catch that error.